beacon-chain/rpc/validator/proposer.go

```diff
@@ -246,51 +249,66 @@ func (vs *Server) eth1DataMajorityVote(ctx context.Context, beaconState *stateTr
+	timeOfLastBlockByEarliestValidTime, err := vs.Eth1BlockFetcher.BlockTimeByHeight(ctx, lastBlockByEarliestValidTime)
 	if err != nil {
-		log.WithError(err).Error("Failed to get block number for previous voting period")
+		log.WithError(err).Error("Failed to get time of last block by earliest valid time")
 		return vs.randomETH1DataVote(ctx)
 	}
-	eth1FollowDistance := int64(params.BeaconConfig().Eth1FollowDistance)
-	currentPeriodInitialBlock := big.NewInt(0).Sub(currentPeriodBlockNumber, big.NewInt(eth1FollowDistance))
-	previousPeriodInitialBlock := big.NewInt(0).Sub(previousPeriodBlockNumber, big.NewInt(eth1FollowDistance))
+	// Increment the earliest block if the original block's time is before valid time.
+	// This is very likely to happen because BlockTimeByHeight returns the last block AT OR BEFORE the specified time.
+	if timeOfLastBlockByEarliestValidTime < earliestValidTime {
+		lastBlockByEarliestValidTime = big.NewInt(0).Add(lastBlockByEarliestValidTime, big.NewInt(1))
+	}
 
-	currentDepositCount, _ := vs.DepositFetcher.DepositsNumberAndRootAtHeight(ctx, currentPeriodInitialBlock)
-	if currentDepositCount == 0 {
-		return vs.ChainStartFetcher.ChainStartEth1Data(), nil
+	lastBlockByLatestValidTime, err := vs.Eth1BlockFetcher.BlockNumberByTimestamp(ctx, latestValidTime)
+	if err != nil {
+		log.WithError(err).Error("Failed to get last block by latest valid time")
+		return vs.randomETH1DataVote(ctx)
+	}
+	timeOfLastBlockByLatestValidTime, err := vs.Eth1BlockFetcher.BlockTimeByHeight(ctx, lastBlockByLatestValidTime)
+	if err != nil {
+		log.WithError(err).Error("Failed to get time of last block by latest valid time")
+		return vs.randomETH1DataVote(ctx)
 	}
 
-	if len(beaconState.Eth1DataVotes()) == 0 {
-		eth1Data, err := vs.defaultEth1DataResponse(ctx, currentPeriodBlockNumber)
-		if err != nil {
-			log.WithError(err).Error("Failed to get eth1 data from current period block number")
-			return vs.randomETH1DataVote(ctx)
-		}
-		return eth1Data, nil
+	lastBlockDepositCount, lastBlockDepositRoot := vs.DepositFetcher.DepositsNumberAndRootAtHeight(ctx, lastBlockByLatestValidTime)
+	if lastBlockDepositCount == 0 {
+		return vs.ChainStartFetcher.ChainStartEth1Data(), nil
 	}
 
-	inRangeVotes, err := vs.inRangeVotes(ctx, beaconState, currentPeriodInitialBlock, previousPeriodInitialBlock)
+	inRangeVotes, err := vs.inRangeVotes(ctx, beaconState, lastBlockByEarliestValidTime, lastBlockByLatestValidTime)
 	if err != nil {
 		return nil, err
 	}
 	if len(inRangeVotes) == 0 {
-		eth1Data, err := vs.defaultEth1DataResponse(ctx, currentPeriodBlockNumber)
-		if err != nil {
-			log.WithError(err).Error("Failed to get eth1 data from current period block number")
-			return vs.randomETH1DataVote(ctx)
+		// If latest block is in range and it doesn't undo deposits, then choose that block.
+		// Otherwise take the current eth1data.
+		currentETH1Data := vs.HeadFetcher.HeadETH1Data()
+		if timeOfLastBlockByLatestValidTime >= earliestValidTime {
```

Comment:
Isn't this already a given ?
```go
	earliestValidTime := votingPeriodStartTime - 2*params.BeaconConfig().SecondsPerETH1Block*uint64(eth1FollowDistance)
	latestValidTime := votingPeriodStartTime - params.BeaconConfig().SecondsPerETH1Block*uint64(eth1FollowDistance)
```
Unless the assumption is `timeOfLastBlockByLatestValidTime` untrusted, then the check should be shifted above when we first retrieve it in `BlockNumberByTimestamp`